apps/webapp/src/components/staking/account/delegation-value-view/staking-actions/form-dialog.tsx

```diff
@@ -0,0 +1,114 @@
+import { getIdentityKey } from '@penumbra-zone/types';
+import {
+  Dialog,
+  DialogContent,
+  DialogHeader,
+  IdentityKeyComponent,
+  Input,
+  Button,
+} from '@penumbra-zone/ui';
+import { ValueViewComponent } from '@penumbra-zone/ui/components/ui/tx/view/value';
+import { InputBlock } from '../../../../shared/input-block';
+import { Validator } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/component/stake/v1/stake_pb';
+import { ValueView } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/asset/v1/asset_pb';
+import { FormEvent } from 'react';
+
+const getDialogHeader = (action: 'delegate' | 'undelegate', validator: Validator) => {
+  if (action === 'delegate') return `Delegate to ${validator.name}`;
+  return `Undelegate from ${validator.name}`;
+};
+
+const getCapitalizedAction = (action: 'delegate' | 'undelegate') =>
+  action.replace(/^./, firstCharacter => firstCharacter.toLocaleUpperCase());
+
+/**
+ * Renders a dialog with a form for delegating to, or undelegating from, a
+ * validator.
+ */
+export const FormDialog = ({
+  action,
+  validator,
+  amount,
+  delegationTokens,
+  unstakedTokens,
+  onChangeAmount,
+  onClose,
+  onSubmit,
+}: {
+  /** When defined, the dialog will be open. */
+  action?: 'delegate' | 'undelegate';
+  /** The validator we're delegating to or undelegating from. */
+  validator: Validator;
+  amount: string;
+  /**
+   * A `ValueView` representing the address's balance of delegation tokens. Used
+   * to show the user how many tokens they have available to undelegate.
+   */
+  delegationTokens: ValueView;
+  /**
+   * A `ValueView` representing the address's balance of staking (UM) tokens.
+   * Used to show the user how many tokens they have available to delegate.
+   */
+  unstakedTokens?: ValueView;
+  onChangeAmount: (amount: string) => void;
+  onClose: () => void;
+  onSubmit: () => void;
+}) => {
+  const handleOpenChange = (open: boolean) => {
+    if (!open) onClose();
+  };
+
+  const handleSubmit = (e: FormEvent) => {
+    e.preventDefault();
+    onSubmit();
+  };
+
+  return (
+    <Dialog open={!!action} onOpenChange={handleOpenChange}>
+      <DialogContent size='sm'>
+        {!!action && (
+          <>
+            <DialogHeader>{getDialogHeader(action, validator)}</DialogHeader>
+            <form className='flex flex-col gap-4 overflow-hidden px-4 pb-4' onSubmit={handleSubmit}>
+              <div className='flex flex-col'>
+                <div className='truncate'>{validator.name}</div>
```

Comment:
I feel like this is too long for the dialog title. Perhaps this should be large/bold in the body.

Reply:
Did you mean to make this comment a few lines above, under the `<DialogHeader />`? 

I intentionally repeated the validator name in the body since it can be too long for the dialog header. That said, I could definitely shorten the header to just say "Delegate" or "Undelegate", rather than "Delegate _to_ [ValidatorName]"...?

Reply:
<img width="474" alt="image" src="https://github.com/penumbra-zone/web/assets/1121544/f9cdfeca-7129-4fdf-879c-2452560af746">

For now, I'm shortening it to just the action per your comment. We can revisit it if needed.